ocaml/typing/solver.ml

```diff
@@ -635,24 +648,45 @@ module Solver_mono (C : Lattices_mono) = struct
     print_raw obj ~verbose ppf
       (match check_const obj m with None -> m | Some a -> Amode a)
 
-  let newvar_above (type a) (obj : a C.obj) = function
-    | Amode a when C.le obj (C.max obj) a -> Amode a, false
-    | m ->
-      let m' = newvar obj in
-      let r = submode_try ~logging:false obj m m' in
-      (match r with Ok None -> () | _ -> assert false);
-      allow_right m', true
-
-  let newvar_below (type a) (obj : a C.obj) = function
-    | Amode a when C.le obj a (C.min obj) -> Amode a, false
-    | m ->
-      let m' = newvar obj in
-      let r = submode_try ~logging:false obj m' m in
-      (match r with
-      | Ok None -> ()
-      | Ok (Some _) -> assert false
-      | Error _ -> assert false);
-      allow_left m', true
+  let newvar obj = Amodevar (Amorphvar (fresh obj, C.id))
+
+  let newvar_above (type a r) (obj : a C.obj) (m : (a, allowed * r) mode) =
+    match disallow_right m with
+    | Amode a ->
+      if C.le obj (C.max obj) a
+      then Amode a, false
+      else Amodevar (Amorphvar (fresh ~lower:a obj, C.id)), true
+    | Amodevar mv ->
+      (* [~lower] is not precise (because [mlower mv] is not precise), but
+         it doesn't need to be *)
+      ( Amodevar
+          (Amorphvar (fresh ~lower:(mlower obj mv) ~vlower:[mv] obj, C.id)),
+        true )
+    | Amodejoin (a, mvs) ->
+      (* [~lower] is not precise here, but it doesn't need to be *)
+      Amodevar (Amorphvar (fresh ~lower:a ~vlower:mvs obj, C.id)), true
+
+  let newvar_below (type a l) (obj : a C.obj) (m : (a, l * allowed) mode) =
+    match disallow_left m with
+    | Amode a ->
+      if C.le obj a (C.min obj)
+      then Amode a, false
+      else Amodevar (Amorphvar (fresh ~upper:a obj, C.id)), true
+    | Amodevar mv ->
+      let u = fresh obj in
+      let mu = Amorphvar (u, C.id) in
+      let mu' = Amorphvar (u, C.id) in
```

Comment:
Why make both `mu` and `mu'`? `Amorphvar` isn't mutable. I think this is unnecessary.

Reply:
It was a workaround of some allowed/disallowed problem, to which I have now found a better solution.